lib/string_decoder.js

```diff
@@ -24,52 +24,121 @@
 const Buffer = require('buffer').Buffer;
 const internalUtil = require('internal/util');
 const isEncoding = Buffer[internalUtil.kIsEncodingSymbol];
+const {
+  copy, latin1Slice, asciiSlice, hexSlice, utf8Slice, ucs2Slice, base64Slice
+} = process.binding('buffer');
 
-// Do not cache `Buffer.isEncoding` when checking encoding names as some
-// modules monkey-patch it to support additional encodings
-function normalizeEncoding(enc) {
-  const nenc = internalUtil.normalizeEncoding(enc);
-  if (typeof nenc !== 'string' &&
-      (Buffer.isEncoding === isEncoding || !Buffer.isEncoding(enc)))
-    throw new Error(`Unknown encoding: ${enc}`);
-  return nenc || enc;
+const encodings = [
+  // 0
+  [
+    'utf8', // normalized encoding name string
+    4, // buffer size
+    (self) => { self.fillLast = utf8FillLast; } // StringDecoder initialization
+  ],
+  // 1
+  [
+    'utf16le',
+    4,
+    (self) => { self.text = utf16Text; self.end = utf16End; }
+  ],
+  // 2
+  [
+    'latin1',
+    0,
+    (self) => { self.text = latin1Text; self.end = simpleEnd; }
+  ],
+  // 3
+  [
+    'base64',
+    3,
+    (self) => { self.text = base64Text; self.end = base64End; }
+  ],
+  // 4
+  [
+    'ascii',
+    0,
+    (self) => { self.text = asciiText; self.end = simpleEnd; }
+  ],
+  // 5
+  [
+    'hex',
+    0,
+    (self) => { self.text = hexText; self.end = simpleEnd; }
+  ]
+];
+
+function translateEncoding(enc) {
+  if (!enc) return 0;
```

Comment:
Would it affect performance if we use constants with names instead of number literals for indices?